keps/sig-api-machinery/20190324-declatative-field-manager.md

```diff
@@ -0,0 +1,177 @@
+
+### User Stories [optional]
+
+#### Git Managed Configuration
+
+Alice has a repo containing her current set of Kubernetes manifests as yaml in git.
+The repo gets managed using Gitflow.
+Master branch is automatically applied to the production Kubernetes cluster by a push triggered Jenkins job.
+Staging branch is automatically applied to the staging Kubernetes cluster by a push triggered Jenkins job.
+All members of her team file pull requests against the repo to bring changes into staging and production.
+
+Aside from that, Alice and her team got services that directly manage their objects, like controllers and webhooks,
+as well as services that updated labels for some objects.
+
+They want to make use of the new Server Side Apply, so concurrent changes to objects from different actors are possible without unwanted overlap/conflicts.
+
+To do so, assuming the changes in this KEP, Alice would add the newly introduced field to all manifests in their config management repo. The field would be set to e.g. `configManagement` (or their repo name).
+Additionally, she would update the manifests sent by different actors (like controllers) to contain the field with a value like `controllerName`. Note that this might already be the case for some through the defaulting to the user-agent inside the apply and update pathway.
+
+If now the team updates their repo with a change to their manifests, Jenkins again applies them as usual. On the apiserver side though, the fieldManager set in the applied manifest gets used and all fields set now get owned by the `configManagement` fieldManager.
+
+Again as usual other actors step in to change the objects. They as well use their fieldManager value either set in the manifest file they apply, or based on their user-agent/configuration.
+One of those actors is setting annotations for every service with external information.
+As the actor now only updates annotations not already owned by the `configManagement` fieldManager, applying them will cause the actor to own them.
+
+All actors now can coexist and manage their set of fields without interfering with each other.
+
+### Risks and Mitigations
+
+One risk might be, that introducing this change would mean there is a field that does not get persisted to storage.
+As a result, setting it results in different behavior on the apiserver, but does not reflect when reding the object back from the apiserver.
+
+This can seem unintuitive and we need to make sure documentation is right on this.
+
+## Design Details
+
+### Test Plan
+
+e2e and integration tests should be added accordingly.
+
+### Graduation Criteria
+
+TBD
+
+### Upgrade / Downgrade Strategy
+
+No fields get persisted.
+When the field is not set due to an old client, the current Server Side Apply defaulting is the fallback.
+When the field is set by a new client, it will get ignored by an old apiserver.
```

Comment:
Is this bad?
If it is an annotation and gets set by a new client to an old apiserver, the field won't be excluded from managedFields as well. Which would mean, on an upgrade we might have to handle this.

Reply:
Can you make the proposed API explicit in the KEP? It sounds like this is suggesting a write-only, non-persisted field to hold the manager name.

How would that field be described in schema docs?
What would the behavior be when sent for non-apply operations like create/update?

Reply:
Thanks for asking @liggitt .
I'm currently starting to work on a first implementation to verify the idea and also be able to answer your questions regarding the schema docs. Have to think about the exact wording.

But yes, the field will be write-only and non-persisted.
It is just for passing information to the apiserver.
I'll update the KEP with more details.

So far i'm favoring the field to be in ObjectMeta so it will look like:
```yaml
metadata:
  name: foo
  namespace: foo
  manager: jenkins
  managedFields: ...
```
If i remember correctly, for non-apply setting the manager is optional right now and the user-agent is the defaulting combined with timestamping etc.
So if the field is set for non-apply it should behave just like it does now and override the user-agent (but, for both operations, get overwritten by the api option).

But i'll add this in more detail to the kep. Already wanted to do it, should finally have time on the weekend.

If you already got questions/feedback i'm happy to hear it.

Reply:
@liggitt would you like to take another look?
The current implementation can be found here: https://github.com/kubernetes/kubernetes/pull/75917

Reply:
Sorry for the slow response, I did not see this until ~yesterday